README.md
![main image](https://user-images.githubusercontent.com/14051960/144635813-82670c30-44c0-43e6-ad61-1505d5b8d929.png)

[![codecov](https://codecov.io/gh/horprogs/Just-validate/branch/master/graph/badge.svg?token=O6DXXL5TUU)](https://codecov.io/gh/horprogs/Just-validate) <a href="https://bundlephobia.com/result?p=just-validate@latest" target="\_parent"> <img alt="" src="https://badgen.net/bundlephobia/minzip/just-validate@latest" /> </a>
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/6c7a25cc9fdb4bf8869884339418352d)](https://www.codacy.com/gh/horprogs/Just-validate/dashboard?utm_source=github.com\&utm_medium=referral\&utm_content=horprogs/Just-validate\&utm_campaign=Badge_Grade)
[![Known Vulnerabilities](https://snyk.io/test/github/horprogs/Just-validate/badge.svg)](https://snyk.io/test/github/horprogs/Just-validate)
[![Release workflow](https://github.com/horprogs/Just-validate/workflows/Test%20and%20Release/badge.svg)](https://github.com/horprogs/Just-validate/actions)

Modern, simple, lightweight (~5kb gzip) form validation library written in Typescript, with no dependencies (no JQuery!).
Support a wide range of predefined rules, async, files, dates validation, custom error messages and styles, localization.
Supporting writing custom rules and plugins.

***

**NOTE**

**This is documentation for JustValidate 2. If you are looking for the old version, you could find it [here](https://github.com/horprogs/Just-validate/tree/v1).**

***

## Why JustValidate?

It's a right choice for you, if you have a site, a landing page without React, JQuery etc.
and you want to quick, simple and powerful solution for validating your form.

## Features

*   small size and zero dependencies
*   no need to change your HTML
*   a wide range of pre-defined rules
*   custom rules
*   support plugins
*   custom styles and css classes for invalid fields and error messages
*   custom messages
*   showing tooltips as error messages
*   localization (defining error messages for different languages)
*   user-friendly setup: console warning messages if something incorrect
*   written in Typescript and good test coverage

Please, check out the examples https://just-validate.dev/examples/

## Installation

### npm

```shell
npm install just-validate --save
```

### yarn

```shell
yarn add just-validate
```

And then use it as an imported module:

```js
import JustValidate from 'just-validate';

const validate = new JustValidate('#form');
```

Or if you don't use module bundlers, just include JustValidate script on your page from CDN and call it as `window.JustValidate`:

```html
<script src="https://unpkg.com/just-validate@latest/dist/just-validate.production.min.js"></script>
<body>
  <script>
    const validate = new window.JustValidate('#form');
  </script>
</body>
```

## Predefined rules

There are plenty of rules which you could use out of the box:

*   required, non-empty fields
*   valid email address
*   min/max text length
*   valid number
*   min/max number
*   valid password
*   valid strong password
*   check for the custom regexp
*   min/max count of uploaded files
*   min/max size, types, extensions, names of uploaded files
*   format date, check for isAfter/isBefore dates

## Quick start

Let's say we have a basic HTML layout:

```html
<form action="#" id="form" autocomplete="off">
  <label for="name">Enter your name</label>
  <input
    type="text"
    class="form__input form-control"
    placeholder="Enter your name"
    autocomplete="off"
    name="name"
    id="name"
  />
  <label for="email">Enter your email</label>
  <input
    type="email"
    class="form__input form-control"
    placeholder="Enter your email"
    autocomplete="off"
    name="email"
    id="email"
  />
  <button class="btn btn-primary" id="submit-btn">Submit</button>
</form>
```

Next, let's add JustValidate to our layout and define some simple rules.

First, we should create the instance `new JustValidate('#form')` by passing a form selector, or the element as an argument.

Second, we call `.addField()` with a field selector as the first argument and an array of rules as the second argument.

```js
const validation = new JustValidate('#form');

validation
  .addField('#name', [
    {
      rule: 'minLength',
      value: 3,
    },
    {
      rule: 'maxLength',
      value: 30,
    },
  ])
  .addField('#email', [
    {
      rule: 'required',
      errorMessage: 'Email is required',
    },
    {
      rule: 'email',
      errorMessage: 'Email is invalid!',
    },
  ]);
```

And that's it! Now our form is validated!

## More fields, group fields and rules

Let's check more advanced example:

```html
<form action="#" id="form" autocomplete="off">
  <div class="row">
    <div class="col">
      <label for="name">Enter your name</label>
      <input
        type="text"
        class="form__input form-control"
        placeholder="Enter your name"
        autocomplete="off"
        name="name"
        id="name"
      />
    </div>
    <div class="col">
      <label for="email">Enter your email</label>
      <input
        type="email"
        class="form__input form-control"
        placeholder="Enter your email"
        autocomplete="off"
        name="email"
        id="email"
      />
    </div>
  </div>
  <div class="form-group mt-3">
    <label for="password">Enter your password</label>
    <input
      type="password"
      class="form__input form-control"
      placeholder="Enter your password"
      autocomplete="off"
      name="password"
      id="password"
    />
  </div>
  <div class="form-group mt-3">
    <label for="password">Repeat your password</label>
    <input
      type="password"
      class="form__input form-control"
      placeholder="Repeat your password"
      autocomplete="off"
      name="repeat-password"
      id="repeat-password"
    />
  </div>
  <div class="form-group mt-3">
    <label for="password">Enter your message</label>
    <textarea
      name="msg"
      cols="30"
      rows="10"
      class="form__textarea form-control"
      id="message"
    ></textarea>
  </div>
  <div class="form-group mt-4">
    <label for="favorite_animal_select" class="form-label"
      >Select you favorite animal</label
    >
    <select name="pets" id="favorite_animal_select" class="form-select">
      <option value="">--Please choose an option--</option>
      <option value="dog">Dog</option>
      <option value="cat">Cat</option>
      <option value="hamster">Hamster</option>
      <option value="parrot">Parrot</option>
      <option value="spider">Spider</option>
      <option value="goldfish">Goldfish</option>
    </select>
  </div>

  <div class="form-group mt-4">
    <div class="form-check">
      <label class="form-check-label" for="consent_checkbox"
        >I agree to provide the information</label
      >
      <input type="checkbox" id="consent_checkbox" class="form-check-input" />
    </div>
  </div>
  <div
    class="form-group mt-4"
    id="read_terms_checkbox_group"
    style="width: 250px"
  >
    <div class="form-check">
      <label class="form-check-label" for="read_terms_checkbox_group_1"
        >I have read Privacy Policy</label
      >
      <input
        type="checkbox"
        name="checkbox-group-fruit"
        id="read_terms_checkbox_group_1"
        class="form-check-input"
      />
    </div>
    <div class="form-check">
      <label class="form-check-label" for="read_terms_checkbox_group_2"
        >I have read Terms Of Use</label
      >
      <input
        type="checkbox"
        name="checkbox-group-fruit"
        id="read_terms_checkbox_group_2"
        class="form-check-input"
      />
    </div>
    <div class="form-check">
      <label class="form-check-label" for="read_terms_checkbox_group_3"
        >I have read Cookies Policy</label
      >
      <input
        type="checkbox"
        name="checkbox-group-fruit"
        id="read_terms_checkbox_group_3"
        class="form-check-input"
      />
    </div>
  </div>
  <div class="mt-4 form-group">
    <div class="pb-1">Please select the preferred way for communication</div>
    <div
      class="form-check"
      id="communication_radio_group"
      style="max-width: 200px"
    >
      <input
        type="radio"
        name="radio"
        class="form-check-input"
        id="communication_radio_group_1"
      />
      <label class="form-check-label" for="communication_radio_group_1">
        Email
      </label>
      <br />
      <input
        type="radio"
        name="radio"
        class="form-check-input"
        id="communication_radio_group_2"
      />
      <label class="form-check-label" for="communication_radio_group_2">
        SMS
      </label>
    </div>
  </div>

  <div class="d-grid mt-4">
    <button class="btn btn-primary" id="submit-btn">Submit</button>
  </div>
</form>
```

```js
const validation = new JustValidate('#form', {
  errorFieldCssClass: 'is-invalid',
});

validation
  .addField('#name', [
    {
      rule: 'minLength',
      value: 3,
    },
    {
      rule: 'maxLength',
      value: 30,
    },
  ])
  .addField('#email', [
    {
      rule: 'required',
      errorMessage: 'Field is required',
    },
    {
      rule: 'email',
      errorMessage: 'Email is invalid!',
    },
  ])
  .addField('#password', [
    {
      rule: 'password',
    },
  ])
  .addField('#message', [
    {
      validator: (value) => {
        return value[0] === '!';
      },
    },
  ])
  .addField('#consent_checkbox', [
    {
      rule: 'required',
    },
  ])
  .addField('#favorite_animal_select', [
    {
      rule: 'required',
    },
  ])
  .addRequiredGroup(
    '#read_terms_checkbox_group',
    'You should select at least one communication channel'
  )
  .addRequiredGroup('#communication_radio_group')
  .onSuccess((event) => {
    console.log('Validation passes and form submitted', event);
  });
```

## Instance setting

    new JustValidate(
       form: string | Element,
       globalConfig?: {
         errorFieldStyle: Partial<CSSStyleDeclaration>,
         errorFieldCssClass: string,
         errorLabelStyle: Partial<CSSStyleDeclaration>,
         errorLabelCssClass: string,
         lockForm: boolean,
         testingMode: boolean,
         focusInvalidField?: boolean,
         tooltip?: {
           position: 'left' | 'top' | 'right' | 'bottom',
         },
       },
       dictLocale?: {
         key: string;
         dict: {
           [localeKey: string]: string,
         };
       }[];
    );

For example, full setting:

```js
const validation = new JustValidate(
  '#form',
  {
    errorFieldCssClass: 'is-invalid',
    errorFieldStyle: {
      border: '1px solid red',
    },
    errorLabelCssClass: 'is-label-invalid',
    errorLabelStyle: {
      color: 'red',
      textDecoration: 'underlined',
    },
    focusInvalidField: true,
    lockForm: true,
    tooltip: {
      position: 'top',
    },
  },
  [
    {
      key: 'Name is too short',
      dict: {
        ru: 'Имя слишком короткое',
        es: 'El nombre es muy corto',
      },
    },
    {
      key: 'Field is required',
      dict: {
        ru: 'Обязательное поле',
        es: 'Se requiere campo',
      },
    },
  ]
);
```

*   `lockForm` - if true, lock form during validation.
    Could be useful for async validation to make it impossible for user to interact with the form
*   `focusInvalidField` - if true, the first invalid field will be focused after the form submitting
*   `tooltip` if the field defined, tooltips will be displayed instead of regular error labels.
    It has `position` field which could be `'left' | 'top' | 'right' | 'bottom'`
*   localization object will be explained in Localization section

### Methods

Define validation rules for the new field:

    .addField(
      field: string,
      rules: {
        rule?: Rules;
        errorMessage?: string | (
          value: string | boolean,
          context: FieldsInterface
        ) => string);
        validator?: (
          value: string | boolean,
          context: FieldsInterface
        ) => boolean | (() => Promise<boolean>);
        value?: number | string | RegExp;
      },
      config?: {
        errorFieldStyle: Partial<CSSStyleDeclaration>;
        errorFieldCssClass: string;
        errorLabelStyle: Partial<CSSStyleDeclaration>;
        errorLabelCssClass: string;
        successFieldStyle?: Partial<CSSStyleDeclaration>;
        successFieldCssClass: string;
        successLabelStyle?: Partial<CSSStyleDeclaration>;
        successLabelCssClass: string;
        tooltip?: {
          position: 'left' | 'top' | 'right' | 'bottom';
        };
        successMessage?: string;
      }
    )

Make the new group field required. It could be a group of checkboxes or radio buttons.

It means that at least one input in the group should be checked/selected.

    .addRequiredGroup(
        groupField: string,
        errorMessage?: string,
        config?: {
          errorFieldStyle: Partial<CSSStyleDeclaration>;
          errorFieldCssClass: string;
          errorLabelStyle: Partial<CSSStyleDeclaration>;
          errorLabelCssClass: string;
          successFieldStyle?: Partial<CSSStyleDeclaration>;
          successFieldCssClass: string;
          successLabelStyle?: Partial<CSSStyleDeclaration>;
          successLabelCssClass: string;
          tooltip?: {
            position: 'left' | 'top' | 'right' | 'bottom';
          };
        },
        successMessage?: string,
    )

`.onSuccess(event)` - callback if validation success

`.onFail(fields)` - callback if validation failed

If you define `successMessage` argument, this text will be shown if validation (all rules) passes.
You could customize it by defining `successFieldStyle`, `successFieldCssClass`, `successLabelStyle`, `successLabelCssClass`
as we do for error messages.

## Rule object

    {
      rule: string,
      errorMessage?: string | (
          value: string | boolean,
          context: FieldsInterface
        ) => string);
      validator: (
        value: string | boolean,
        context: object
     ) => Boolean | Promise;
     value: number | string;
    }

Field `rule` could be one of these values:

*   `required` - Required field, not empty
*   `email` - Valid email address
*   `minLength` - Limit the minimum text length
*   `maxLength` - Limit the maximum text length
*   `number` - Value should be a number
*   `minNumber` - Number should be more than defined value
*   `maxNumber` - Number should be less than defined value
*   `password` - Minimum eight characters, at least one letter and one number
*   `strongPassword` - Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
*   `customRegexp` - Custom regexp expression
*   `minFilesCount` - Uploaded files count should be more than defined value
*   `maxFilesCount` - Uploaded files count should be less than defined value
*   `files` - Uploaded files attributes should be valid, based on `value` config (check details below in Files validation section)

Field `errorMessage` used for customizing the error message for this rule. It could be a string, or a function which returns a string
(it helps to create conditional messages, based on the field value).

Field `value` used only with `minLength`, `maxLength`, `minNumber`, `maxNumber`, `customRegexp` rules to define the rule value, for example:

```js
validation.addField('#name', [
  {
    rule: 'minLength',
    value: 3,
  },
]);
```

```js
validation.addField('#password', [
  {
    rule: 'customRegexp',
    value: /(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[@$!%*#?&^_-]).{8,}/,
  },
]);
```

Field `validator` used for custom validation. It should return a boolean or a function which returns a Promise.

```js
validation.addField('#text', [
  {
    validator: (value) => value[value.length - 1] === '.',
  },
]);
```

```js
validation.addField('#email', [
  {
    validator: (value) => () => {
      return new Promise((resolve) => {
        setTimeout(() => {
          resolve(true);
        }, 500);
      });
    },
  },
]);
```

Also, it's possible to do validation based on other fields:

```js
validation.addField('#repeat-password', [
  {
    validator: (value, fields) => {
      if (fields['#password'] && fields['#password'].elem) {
        const repeatPasswordValue = fields['#password'].elem.value;

        return value === repeatPasswordValue;
      }

      return true;
    },
    errorMessage: 'Passwords should be the same',
  },
]);
```

## Files validation

It's possible to validate min/max count of uploaded files and check for type, extension, min/max size and name of uploaded files.

### Uploaded files count

```html
<input type="file" name="file" id="file" multiple />
```

```js
validation.addField('#file', [
  {
    rule: 'minFilesCount',
    value: 1,
  },
  {
    rule: 'maxFilesCount',
    value: 3,
  },
]);
```

### Files attributes validation

```js
validation.addField('#file', [
  {
    rule: 'files',
    value: {
      files: {
        extensions: ['jpeg', 'png'],
        maxSize: 25000,
        minSize: 1000,
        types: ['image/jpeg', 'image/png'],
        names: ['file1.jpeg', 'file2.png'],
      },
    },
  },
]);
```

Format of value for `files` rule:

```
{
  files: {
    extensions?: string[];
    types?: string[];
    minSize?: number;
    maxSize?: number;
    names?: string[];
  }
}

```

`minSize` and `maxSize` should be defined in bytes.

### Date validation

Date validation is performed by a separate plugin [JustValidatePluginDate](https://github.com/horprogs/just-validate-plugin-date) to avoid extending the library size. The plugin uses [date-fns](https://github.com/date-fns/date-fns) for operating with dates.

It is possible to validate for defined format, isAfter, isBefore dates. It works both for text input and for date input.

#### Configuration

`JustValidatePluginDate` takes 1 argument: function which returns config object:

```js
JustValidatePluginDate((fields) => ({
  required: boolean,
  format: string,
  isBefore: Date | string,
  isAfter: Date | string,
}));
```

If it's not required, validation will pass if the value is empty.

It is useful to use `fields` for accessing other fields and their value to set values in the config.

#### Validate for format

```js
import JustValidatePluginDate from 'just-validate-plugin-date';

validation.addField('#date', [
  {
    plugin: JustValidatePluginDate(() => ({
      format: 'dd MMM yyyy',
    })),
    errorMessage: 'Date should be in dd MMM yyyy format (e.g. 20 Dec 2021)',
  },
]);
```

#### Validate for isBefore/isAfter

To be able to validate isBefore/isAfter it's important to define **the same format which supposed to use in isBefore/isAfter**,
otherwise the library will not be able to parse this date string correctly.

```js
import JustValidatePluginDate from 'just-validate-plugin-date';

validation.addField('#date', [
  {
    plugin: JustValidatePluginDate(() => ({
      format: 'dd/MM/yyyy',
      isBefore: '15/12/2021',
      isAfter: '10/12/2021',
    })),
    errorMessage: 'Date should be between 10/12/2021 and 15/12/2021',
  },
]);
```

Also, you could perform validation depends on other fields:

```js
import JustValidatePluginDate from 'just-validate-plugin-date';

validation
  .addField('#date-start', [
    {
      plugin: JustValidatePluginDate(() => ({
        format: 'dd/MM/yyyy',
      })),
      errorMessage: 'Date should be in dd/MM/yyyy format (e.g. 15/10/2021)',
    },
  ])
  .addField('#date-between', [
    {
      plugin: JustValidatePluginDate(() => ({
        format: 'dd/MM/yyyy',
      })),
      errorMessage: 'Date should be in dd/MM/yyyy format (e.g. 15/10/2021)',
    },
    {
      plugin: JustValidatePluginDate((fields) => ({
        format: 'dd/MM/yyyy',
        isAfter: fields['#date-start'].elem.value,
        isBefore: fields['#date-end'].elem.value,
      })),
      errorMessage: 'Date should be between start and end dates',
    },
  ])
  .addField('#date-between-required', [
    {
      plugin: JustValidatePluginDate(() => ({
        required: true,
        format: 'dd/MM/yyyy',
      })),
      errorMessage: 'Date should be in dd/MM/yyyy format (e.g. 15/10/2021)',
    },
    {
      plugin: JustValidatePluginDate((fields) => ({
        required: true,
        format: 'dd/MM/yyyy',
        isAfter: fields['#date-start'].elem.value,
        isBefore: fields['#date-end'].elem.value,
      })),
      errorMessage: 'Date should be between start and end dates',
    },
  ])
  .addField('#date-end', [
    {
      plugin: JustValidatePluginDate(() => ({
        format: 'dd/MM/yyyy',
      })),
      errorMessage: 'Date should be in dd/MM/yyyy format (e.g. 15/10/2021)',
    },
  ]);
```

#### Using with date input

If you use date input, you don't need to define `format` field, because it's always in `yyyy-mm-dd` format,
you just should define isBefore/isAfter:

```js
validation.addField('#date-between', [
  {
    plugin: JustValidatePluginDate((fields) => ({
      isAfter: document.querySelector('#date-start').value,
      isBefore: document.querySelector('#date-end').value,
    })),
    errorMessage: 'Date should be between start and end dates',
  },
]);
```

### Localization

You could define your own translations for different languages. To do that you should define `dictLocale` array, like:

      [
        {
          key: 'Name is too short',
          dict: {
            ru: 'Имя слишком короткое',
            es: 'El nombre es muy corto',
          },
        },
        {
          key: 'Field is required',
          dict: {
            ru: 'Обязательное поле',
            es: 'Se requiere campo',
          },
        },
      ]

Field `key` should be defined as a key string, which also should be defined as `errorMessage` in a rule object.

`dict` should be an object with languages keys with their translations.

To switch a language you should call `validation.setCurrentLocale('ru');`.
The argument for `setCurrentLocale()` method you should pass the key, which you defined in `dict` object,
or you could call with empty argument to set the default language (strings defined in `key` fields).

```js
document.querySelector('#change-lang-btn-en').addEventListener('click', () => {
  validation.setCurrentLocale();
});

document.querySelector('#change-lang-btn-ru').addEventListener('click', () => {
  validation.setCurrentLocale('ru');
});

document.querySelector('#change-lang-btn-es').addEventListener('click', () => {
  validation.setCurrentLocale('es');
});
```

### Plugins

If you need more custom functionality it is possible to write your own plugin. For now there is one official plugin:

*   [JustValidatePluginDate](https://github.com/horprogs/just-validate-plugin-date)


CHANGELOG.md
# [3.3.0](https://github.com/horprogs/Just-validate/compare/v3.2.2...v3.3.0) (2022-01-10)


### Features

* [#35](https://github.com/horprogs/Just-validate/issues/35) add conditional messages ([914d4d2](https://github.com/horprogs/Just-validate/commit/914d4d2ac59d4bcbf7e9e45f03f27cce044b1d32))
* add success messages ([bbf0d82](https://github.com/horprogs/Just-validate/commit/bbf0d829276ca6f31f2487d0380cf29c16bd8ca2))

## [3.2.2](https://github.com/horprogs/Just-validate/compare/v3.2.1...v3.2.2) (2021-12-31)


### Bug Fixes

* [#34](https://github.com/horprogs/Just-validate/issues/34) fix empty value for customRegexp rule ([73cd1d7](https://github.com/horprogs/Just-validate/commit/73cd1d78db7d83017070a5fe283fc6f519a95cf3))

## [3.2.1](https://github.com/horprogs/Just-validate/compare/v3.2.0...v3.2.1) (2021-12-31)


### Bug Fixes

* [#36](https://github.com/horprogs/Just-validate/issues/36) fix checking ruleValue for 0 ([fc8483e](https://github.com/horprogs/Just-validate/commit/fc8483e764c0c03fbff1b8b309a0ccfbd8cf8986))

# [3.2.0](https://github.com/horprogs/Just-validate/compare/v3.1.2...v3.2.0) (2021-12-30)


### Features

* add date validation and plugin support ([28e8382](https://github.com/horprogs/Just-validate/commit/28e838211573acdcdd0db183883ff685ea35cebb))

## [3.1.2](https://github.com/horprogs/Just-validate/compare/v3.1.1...v3.1.2) (2021-12-29)


### Bug Fixes

* fix README and trigger new patch release ([90b6b1e](https://github.com/horprogs/Just-validate/commit/90b6b1e3499eaa43c4058e0c1b65a8cae818fd73))

## [3.1.1](https://github.com/horprogs/Just-validate/compare/v3.1.0...v3.1.1) (2021-12-27)


### Bug Fixes

* upgrade devDependencies ([8f5cd53](https://github.com/horprogs/Just-validate/commit/8f5cd53384ca66a038f768b4badf7accea40dad5))

# [3.1.0](https://github.com/horprogs/Just-validate/compare/v3.0.0...v3.1.0) (2021-12-27)


### Features

* [#32](https://github.com/horprogs/Just-validate/issues/32) add minFilesCount, maxFilesCount, files rules for files validation ([db956dd](https://github.com/horprogs/Just-validate/commit/db956dde812755e49ca52a5dd79da3d70e20207b))
* [#32](https://github.com/horprogs/Just-validate/issues/32) add minFilesCount, maxFilesCount, files rules for files validation ([5056975](https://github.com/horprogs/Just-validate/commit/50569753848605e3d6848b514acf9791086ce432))

# [3.0.0](https://github.com/horprogs/Just-validate/compare/v2.2.1...v3.0.0) (2021-12-27)


### Features

* [#31](https://github.com/horprogs/Just-validate/issues/31) change required logic ([c7b6c78](https://github.com/horprogs/Just-validate/commit/c7b6c7805961dffb574c8cb40978e7a2bdcc2733))


### BREAKING CHANGES

* rules for maxLength, maxNumber etc. will not raise an error if the field is empty and there is no required rule. To validate for required you should define required rule directly.

## [2.2.1](https://github.com/horprogs/Just-validate/compare/v2.2.0...v2.2.1) (2021-12-27)


### Reverts

* Revert "feat(): change required logic \n\n BREAKING CHANGE: rules for maxLength, maxNumber etc. will not raise an error if the field is empty and there is no required rule. To validate for required you should defined required rule directly." ([a1bce7e](https://github.com/horprogs/Just-validate/commit/a1bce7ed00dbeb1934ee5e3ee060ecefd0013ac6))

# [2.2.0](https://github.com/horprogs/Just-validate/compare/v2.1.0...v2.2.0) (2021-12-27)


### Features

* change required logic \n\n BREAKING CHANGE: rules for maxLength, maxNumber etc. will not raise an error if the field is empty and there is no required rule. To validate for required you should defined required rule directly. ([62561a9](https://github.com/horprogs/Just-validate/commit/62561a99d14f40ff6a4d60f89ae2e6d37041095e))

# [2.1.0](https://github.com/horprogs/Just-validate/compare/v2.0.0...v2.1.0) (2021-12-22)


### Features

* [#30](https://github.com/horprogs/Just-validate/issues/30) add removeField method ([9b09503](https://github.com/horprogs/Just-validate/commit/9b09503089686261436e9264090c99c1fd108985))

# [2.0.0](https://github.com/horprogs/Just-validate/compare/v1.5.0...v2.0.0) (2021-12-07)


### Bug Fixes

* add git plugin ([02edd67](https://github.com/horprogs/Just-validate/commit/02edd67393172396168ae4e94d2430d4d8a033eb))
* adjust semantic-release config ([ed33cd1](https://github.com/horprogs/Just-validate/commit/ed33cd1a81129802c16945e4b703ea1a83146669))
* change invalid import ([d28b4b5](https://github.com/horprogs/Just-validate/commit/d28b4b5b679ba0482e5f4bd1757a7fb6e5cda324))
* fix global config ([be01fd8](https://github.com/horprogs/Just-validate/commit/be01fd8c5d88e29b3d70a856ad2e6c05c7bfcca6))
* fix ts config and package.json ([918bbba](https://github.com/horprogs/Just-validate/commit/918bbbaae7a9b35c34ab0f1633b0c0205ccac544))
* improve switching languages, fix focus field ([b0cd51c](https://github.com/horprogs/Just-validate/commit/b0cd51c3f6322517b31564f426a2d577d6d620ab))
* improve tooltip position calculation on scroll ([2c12854](https://github.com/horprogs/Just-validate/commit/2c128546cb1a01799c632ccf86e8cc4eedcc23ae))
* make field invalid if setting is invalid, add tests ([ede7a66](https://github.com/horprogs/Just-validate/commit/ede7a66e0b397fba539a3fa92d5e37f37f2c611e))
* remove unnecessary dep ([67f5549](https://github.com/horprogs/Just-validate/commit/67f5549e70a4bfc9bba147cf3ccb59a0441c0970))
* split settings for site ([2680add](https://github.com/horprogs/Just-validate/commit/2680adda6ac3796322cd368821f0f5c5e9a220ef))
* test change release automation ([1cdcdd8](https://github.com/horprogs/Just-validate/commit/1cdcdd812f1e97d4dce7c3b16227740c61acc201))
* trigger test release ([cda3c21](https://github.com/horprogs/Just-validate/commit/cda3c215dc3ad34ec3ac3c18174c5f735b9aa384))
* trigger test release ([e9eacb9](https://github.com/horprogs/Just-validate/commit/e9eacb9634dabd55affaeb82468d13a42c78f42b))

src/example.ts
/* eslint-disable no-console */
import JustValidate from './main';
import { Rules } from './modules/interfaces';

const fetch = (time = 1000, func?: () => boolean) =>
  new Promise<boolean>((resolve) => {
    setTimeout(() => {
      resolve(func?.() || false);
    }, time);
  });

const advancedExample = () => {
  const validation = new JustValidate(
    '#form',
    {
      errorFieldCssClass: 'is-invalid',
      errorFieldStyle: {
        border: '1px solid red',
      },
      errorLabelCssClass: 'is-label-invalid',
      errorLabelStyle: {
        color: 'red',
        textDecoration: 'underlined',
      },
      focusInvalidField: true,
      lockForm: true,
    },
    [
      {
        key: 'Name is too short',
        dict: {
          ru: 'Имя слишком короткое',
          es: 'El nombre es muy corto',
        },
      },
      {
        key: 'Field is required',
        dict: {
          ru: 'Обязательное поле',
          es: 'Se requiere campo',
        },
      },
    ]
  );

  validation
    .addField('#name', [
      {
        rule: 'minNumber' as Rules,
        value: 10,
      },
      {
        rule: 'maxNumber' as Rules,
        value: 100,
      },
    ])
    .addField('#email', [
      {
        rule: 'required' as Rules,
        errorMessage: 'Field is required',
      },
      {
        rule: 'email' as Rules,
        errorMessage: 'Email is wrong!',
      },
    ])
    .addField(
      '#password',
      [
        {
          rule: 'required' as Rules,
        },
      ],
      {
        tooltip: {
          position: 'bottom',
        },
      }
    )
    .addField('#repeat-password', [
      {
        validator: (value, fields) => {
          if (fields['#password'] && fields['#password'].elem) {
            const repeatPasswordValue = fields['#password'].elem.value;

            return value === repeatPasswordValue;
          }

          return true;
        },
        errorMessage: 'Passwords should be the same',
      },
    ])
    .addField(
      '#message',
      [
        {
          validator: (value) => {
            return () => fetch(0, () => !!value);
          },
        },
      ],
      {
        tooltip: {
          position: 'bottom',
        },
      }
    )
    .addField(
      '#consent_checkbox',
      [
        {
          rule: 'required' as Rules,
        },
      ],
      {
        tooltip: {
          position: 'left',
        },
      }
    )
    .addField(
      '#favorite_animal_select',
      [
        {
          rule: 'required' as Rules,
        },
      ],
      {
        tooltip: {
          position: 'right',
        },
      }
    )
    .addRequiredGroup(
      '#read_terms_checkbox_group',
      'You should select at least one communication channel',
      {
        tooltip: {
          position: 'right',
        },
      }
    )
    .addRequiredGroup('#communication_radio_group', undefined, {
      tooltip: {
        position: 'right',
      },
    })
    .onSuccess((event) => {
      console.log('Validation passes and form submitted', event);
    });

  document
    .querySelector('#change-lang-btn-en')
    ?.addEventListener('click', () => {
      validation.setCurrentLocale('en');
    });

  document
    .querySelector('#change-lang-btn-ru')
    ?.addEventListener('click', () => {
      validation.setCurrentLocale('ru');
    });
};

const basicExample = () => {
  const validation = new JustValidate('#basic_form', {
    errorFieldCssClass: 'is-invalid',
    errorFieldStyle: {
      border: '1px solid red',
    },
    errorLabelCssClass: 'is-label-invalid',
    errorLabelStyle: {
      color: 'red',
      textDecoration: 'underlined',
    },
    focusInvalidField: true,
    lockForm: true,
  });

  validation
    .addField('#basic_name', [
      {
        rule: 'maxNumber' as Rules,
        value: 10,
      },
    ])
    .onSuccess((ev) => {
      ev?.preventDefault();
      console.log('SUCCESS', ev);
    })
    .onFail((fields) => {
      console.log('FAIL', fields);
    });
  // .addField('#basic_email', [
  //   {
  //     rule: 'required' as Rules,
  //     errorMessage: 'Field is required',
  //   },
  //   {
  //     rule: 'email' as Rules,
  //     errorMessage: 'Email is wrong!',
  //   },
  // ]);
};

const conditionalExample = () => {
  const validation = new JustValidate('#conditional_form', {
    errorFieldCssClass: 'is-invalid',
    errorFieldStyle: {
      border: '1px solid red',
    },
    errorLabelCssClass: 'is-label-invalid',
    errorLabelStyle: {
      color: 'red',
      textDecoration: 'underlined',
    },
    focusInvalidField: true,
  });

  validation.onSuccess((ev) => {
    ev?.preventDefault();
    console.log('SUCCESS', ev);
  });

  document
    .querySelector('#conditional_required_name_checkbox')!
    .addEventListener('click', (e) => {
      if ((e.target as HTMLInputElement).checked) {
        validation.addField('#conditional_name', [
          {
            rule: 'required' as Rules,
          },
        ]);
      } else {
        validation.removeField('#conditional_name');
      }
    });
};

const fileExample = () => {
  const validation = new JustValidate('#file_form', {
    errorFieldCssClass: 'is-invalid',
    errorFieldStyle: {
      border: '1px solid red',
    },
    errorLabelCssClass: 'is-label-invalid',
    errorLabelStyle: {
      color: 'red',
      textDecoration: 'underlined',
    },
    focusInvalidField: true,
    lockForm: true,
  });

  validation
    .addField('#file_minmax', [
      {
        rule: 'minFilesCount' as Rules,
        value: 1,
      },
      {
        rule: 'maxFilesCount' as Rules,
        value: 3,
      },
    ])
    .addField('#file_txt', [
      {
        rule: 'minFilesCount' as Rules,
        value: 1,
      },
      {
        rule: 'maxFilesCount' as Rules,
        value: 1,
      },
      {
        rule: 'files' as Rules,
        value: {
          files: {
            extensions: ['txt'],
          },
        },
      },
    ])
    .addField('#file_advanced', [
      {
        rule: 'files' as Rules,
        value: {
          files: {
            extensions: ['jpeg', 'png'],
            maxSize: 25000,
            minSize: 1000,
            types: ['image/jpeg', 'image/png'],
          },
        },
      },
    ])
    .onSuccess((ev) => {
      ev?.preventDefault();
      console.log('SUCCESS', ev);
    })
    .onFail((fields) => {
      console.log('FAIL', fields);
    });
  // .addField('#basic_email', [
  //   {
  //     rule: 'required' as Rules,
  //     errorMessage: 'Field is required',
  //   },
  //   {
  //     rule: 'email' as Rules,
  //     errorMessage: 'Email is wrong!',
  //   },
  // ]);
};

basicExample();
advancedExample();
conditionalExample();
fileExample();
